packages/candid/src/idl.ts

```diff
@@ -921,8 +921,19 @@ export class OptClass<T> extends ConstructType<[T] | []> {
     switch (safeReadUint8(b)) {
       case 0:
         return [];
-      case 1:
-        return [this._type.decodeValue(b, opt._type)];
+      case 1: {
+        const checkpoint = b.save();
+        try {
+          const v = this._type.decodeValue(b, opt._type)
+          return [v];
+        } catch (e : any) {
+          b.restore(checkpoint);
+          // skip value at wire type (to advance b)
+          const v = opt._type.decodeValue(b, opt._type)
```

Comment:
Why do we decode the value again if we are always returning none?

Reply:
Because we need to skip past the value so the buffer is consumed...

Reply:
The motoko code does this a bit differently. It avoids reparsing the candid by parsing to completion and returning a proper value or sentinel value to indicate coercion failure. But that's harder to integrate here without introducing mistakes.